File: data_baseline/openrouter__qwen__qwen3.7-plus/java/rep3/Main.java
import com.sun.net.httpserver.HttpServer;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpExchange;
import java.net.InetSocketAddress;
import java.util.concurrent.Executors;
import java.util.concurrent.ConcurrentHashMap;
import java.io.IOException;
import java.util.*;
import java.util.regex.*;
import java.nio.charset.StandardCharsets;

public class Main {
    public static void main(String[] args) {
        int port = 8080;
        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--port") && i + 1 < args.length) {
                port = Integer.parseInt(args[i + 1]);
            }
        }
        try {
            HttpServer server = HttpServer.create(new InetSocketAddress("0.0.0.0", port), 0);
            DataStore store = new DataStore();
            
            server.createContext("/register", new RegisterHandler(store));
            server.createContext("/login", new LoginHandler(store));
            server.createContext("/logout", new LogoutHandler(store));
            server.createContext("/me", new MeHandler(store));
            server.createContext("/password", new PasswordHandler(store));
            server.createContext("/todos", new TodosHandler(store));
            
            server.setExecutor(Executors.newFixedThreadPool(10));
            server.start();
            System.out.println("Server started on port " + port);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

class JsonHelper {
    public static String object(Object... kv) {
        StringBuilder sb = new StringBuilder("{");
        for (int i = 0; i < kv.length; i += 2) {
            if (i > 0) sb.append(",");
            sb.append("\"").append(escape(kv[i].toString())).append("\":");
            Object v = kv[i+1];
            if (v instanceof String) {
                sb.append("\"").append(escape(v.toString())).append("\"");
            } else if (v instanceof Boolean || v instanceof Number) {
                sb.append(v.toString());
            } else if (v == null) {
                sb.append("null");
            }
        }
        sb.append("}");
        return sb.toString();
    }
    
    public static String array(List<String> items) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < items.size(); i++) {
            if (i > 0) sb.append(",");
            sb.append(items.get(i));
        }
        sb.append("]");
        return sb.toString();
    }

    public static String escape(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }
    
    public static Map<String, Object> parse(String json) {
        Map<String, Object> map = new LinkedHashMap<>();
        if (json == null) return map;
        String inner = json.trim();
        if (inner.startsWith("{") && inner.endsWith("}")) {
            inner = inner.substring(1, inner.length() - 1).trim();
        }
        if (inner.isEmpty()) return map;
        
        Pattern p = Pattern.compile("\"([^\"]+)\"\\s*:\\s*(\"(?:[^\"\\\\]|\\\\.)*\"|true|false|null|-?\\d+(?:\\.\\d+)?)");
        Matcher m = p.matcher(inner);
        while (m.find()) {
            String key = m.group(1);
            String val = m.group(2);
            if (val.startsWith("\"") && val.endsWith("\"")) {
                val = val.substring(1, val.length() - 1);
                val = val.replace("\\\\", "\\").replace("\\\"", "\"").replace("\\n", "\n").replace("\\t", "\t").replace("\\r", "\r");
                map.put(key, val);
            } else if (val.equals("true")) {
                map.put(key, true);
            } else if (val.equals("false")) {
                map.put(key, false);
            } else if (val.equals("null")) {
                map.put(key, null);
            } else {
                try {
                    if (val.contains(".")) {
                        map.put(key, Double.parseDouble(val));
                    } else {
                        map.put(key, Long.parseLong(val));
                    }
                } catch (NumberFormatException e) {
                    map.put(key, val);
                }
            }
        }
        return map;
    }
}

class DataStore {
    private int nextUserId = 1;
    private int nextTodoId = 1;
    private final Map<String, User> usersByUsername = new ConcurrentHashMap<>();
    private final Map<Integer, User> usersById = new ConcurrentHashMap<>();
    private final Map<Integer, Todo> todos = new ConcurrentHashMap<>();
    private final Map<String, Integer> sessions = new ConcurrentHashMap<>();

    public synchronized User register(String username, String password) {
        if (usersByUsername.containsKey(username)) {
            return null;
        }
        User user = new User(nextUserId++, username, password);
        usersByUsername.put(username, user);
        usersById.put(user.id, user);
        return user;
    }

    public User login(String username, String password) {
        User user = usersByUsername.get(username);
        if (user != null && user.password.equals(password)) {
            return user;
        }
        return null;
    }

    public String createSession(int userId) {
        String token = UUID.randomUUID().toString().replace("-", "");
        sessions.put(token, userId);
        return token;
    }

    public void invalidateSession(String token) {
        sessions.remove(token);
    }

    public Integer getSessionUser(String token) {
        return sessions.get(token);
    }

    public User getUserById(int id) {
        return usersById.get(id);
    }

    public synchronized Todo createTodo(int userId, String title, String description) {
        String now = getIsoNow();
        Todo todo = new Todo(nextTodoId++, userId, title, description, false, now, now);
        todos.put(todo.id, todo);
        return todo;
    }

    public Todo getTodo(int id) {
        return todos.get(id);
    }

    public List<Todo> getTodosByUser(int userId) {
        List<Todo> result = new ArrayList<>();
        for (Todo todo : todos.values()) {
            if (todo.userId == userId) {
                result.add(todo);
            }
        }
        result.sort(Comparator.comparingInt(t -> t.id));
        return result;
    }

    public synchronized Todo updateTodo(int id, int userId, String title, String description, Boolean completed) {
        Todo todo = todos.get(id);
        if (todo != null && todo.userId == userId) {
            if (title != null) todo.title = title;
            if (description != null) todo.description = description;
            if (completed != null) todo.completed = completed;
            todo.updatedAt = getIsoNow();
            return todo;
        }
        return null;
    }

    public synchronized boolean deleteTodo(int id, int userId) {
        Todo todo = todos.get(id);
        if (todo != null && todo.userId == userId) {
            todos.remove(id);
            return true;
        }
        return false;
    }

    public boolean checkPassword(int userId, String password) {
        User user = usersById.get(userId);
        return user != null && user.password.equals(password);
    }

    public synchronized void updatePassword(int userId, String newPassword) {
        User user = usersById.get(userId);
        if (user != null) {
            user.password = newPassword;
        }
    }

    private String getIsoNow() {
        return java.time.ZonedDateTime.now(java.time.ZoneOffset.UTC)
                .format(java.time.format.DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'"));
    }
}

class User {
    int id;
    String username;
    String password;
    User(int id, String username, String password) {
        this.id = id;
        this.username = username;
        this.password = password;
    }
}

class Todo {
    int id;
    int userId;
    String title;
    String description;
    boolean completed;
    String createdAt;
    String updatedAt;
    Todo(int id, int userId, String title, String description, boolean completed, String createdAt, String updatedAt) {
        this.id = id;
        this.userId = userId;
        this.title = title;
        this.description = description;
        this.completed = completed;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }
    public String toJson() {
        return JsonHelper.object(
            "id", id,
            "title", title,
            "description", description,
            "completed", completed,
            "created_at", createdAt,
            "updated_at", updatedAt
        );
    }
}

abstract class BaseHandler implements HttpHandler {
    protected final DataStore store;
    protected BaseHandler(DataStore store) {
        this.store = store;
    }

    protected String getCookie(HttpExchange exchange, String name) {
        String cookies = exchange.getRequestHeaders().getFirst("Cookie");
        if (cookies != null) {
            for (String cookie : cookies.split(";")) {
                String[] parts = cookie.trim().split("=", 2);
                if (parts.length == 2 && parts[0].equals(name)) {
                    return parts[1];
                }
            }
        }
        return null;
    }

    protected Integer requireAuth(HttpExchange exchange) throws IOException {
        String token = getCookie(exchange, "session_id");
        if (token == null) {
            sendJson(exchange, 401, JsonHelper.object("error", "Authentication required"));
            return null;
        }
        Integer userId = store.getSessionUser(token);
        if (userId == null) {
            sendJson(exchange, 401, JsonHelper.object("error", "Authentication required"));
            return null;
        }
        return userId;
    }

    protected void sendJson(HttpExchange exchange, int code, String json) throws IOException {
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        byte[] bytes = json.getBytes(StandardCharsets.UTF_8);
        exchange.sendResponseHeaders(code, bytes.length);
        exchange.getResponseBody().write(bytes);
        exchange.getResponseBody().close();
    }
    
    protected void sendNoContent(HttpExchange exchange, int code) throws IOException {
        exchange.sendResponseHeaders(code, -1);
        exchange.getResponseBody().close();
    }

    protected String readBody(HttpExchange exchange) throws IOException {
        java.io.InputStream is = exchange.getRequestBody();
        java.util.Scanner s = new java.util.Scanner(is).useDelimiter("\\A");
        return s.hasNext() ? s.next() : "";
    }
}

class RegisterHandler extends BaseHandler {
    public RegisterHandler(DataStore store) { super(store); }
    public void handle(HttpExchange exchange) throws IOException {
        if (!exchange.getRequestMethod().equals("POST")) {
            sendJson(exchange, 405, JsonHelper.object("error", "Method not allowed"));
            return;
        }
        String body = readBody(exchange);
        Map<String, Object> req = JsonHelper.parse(body);
        Object u = req.get("username");
        Object p = req.get("password");
        
        if (!(u instanceof String) || !((String)u).matches("^[a-zA-Z0-9_]{3,50}$")) {
            sendJson(exchange, 400, JsonHelper.object("error", "Invalid username"));
            return;
        }
        String username = (String) u;
        
        if (!(p instanceof String) || ((String)p).length() < 8) {
            sendJson(exchange, 400, JsonHelper.object("error", "Password too short"));
            return;
        }
        String password = (String) p;
        
        User user = store.register(username, password);
        if (user == null) {
            sendJson(exchange, 409, JsonHelper.object("error", "Username already exists"));
            return;
        }
        sendJson(exchange, 201, JsonHelper.object("id", user.id, "username", user.username));
    }
}

class LoginHandler extends BaseHandler {
    public LoginHandler(DataStore store) { super(store); }
    public void handle(HttpExchange exchange) throws IOException {
        if (!exchange.getRequestMethod().equals("POST")) {
            sendJson(exchange, 405, JsonHelper.object("error", "Method not allowed"));
            return;
        }
        String body = readBody(exchange);
        Map<String, Object> req = JsonHelper.parse(body);
        Object u = req.get("username");
        Object p = req.get("password");
        
        if (u instanceof String && p instanceof String) {
            User user = store.login((String)u, (String)p);
            if (user != null) {
                String token = store.createSession(user.id);
                exchange.getResponseHeaders().add("Set-Cookie", "session_id=" + token + "; Path=/; HttpOnly");
                sendJson(exchange, 200, JsonHelper.object("id", user.id, "username", user.username));
                return;
            }
        }
        sendJson(exchange, 401, JsonHelper.object("error", "Invalid credentials"));
    }
}

class LogoutHandler extends BaseHandler {
    public LogoutHandler(DataStore store) { super(store); }
    public void handle(HttpExchange exchange) throws IOException {
        if (!exchange.getRequestMethod().equals("POST")) {
            sendJson(exchange, 405, JsonHelper.object("error", "Method not allowed"));
            return;
        }
        Integer userId = requireAuth(exchange);
        if (userId == null) return;
        String token = getCookie(exchange, "session_id");
        if (token != null) {
            store.invalidateSession(token);
        }
        sendJson(exchange, 200, JsonHelper.object());
    }
}

class MeHandler extends BaseHandler {
    public MeHandler(DataStore store) { super(store); }
    public void handle(HttpExchange exchange) throws IOException {
        if (!exchange.getRequestMethod().equals("GET")) {
            sendJson(exchange, 405, JsonHelper.object("error", "Method not allowed"));
            return;
        }
        Integer userId = requireAuth(exchange);
        if (userId == null) return;
        User user = store.getUserById(userId);
        sendJson(exchange, 200, JsonHelper.object("id", user.id, "username", user.username));
    }
}

class PasswordHandler extends BaseHandler {
    public PasswordHandler(DataStore store) { super(store); }
    public void handle(HttpExchange exchange) throws IOException {
        if (!exchange.getRequestMethod().equals("PUT")) {
            sendJson(exchange, 405, JsonHelper.object("error", "Method not allowed"));
            return;
        }
        Integer userId = requireAuth(exchange);
        if (userId == null) return;
        String body = readBody(exchange);
        Map<String, Object> req = JsonHelper.parse(body);
        Object oldP = req.get("old_password");
        Object newP = req.get("new_password");
        
        if (!(oldP instanceof String) || !(newP instanceof String)) {
            sendJson(exchange, 400, JsonHelper.object("error", "Invalid request"));
            return;
        }
        
        if (!store.checkPassword(userId, (String)oldP)) {
            sendJson(exchange, 401, JsonHelper.object("error", "Invalid credentials"));
            return;
        }
        
        if (((String)newP).length() < 8) {
            sendJson(exchange, 400, JsonHelper.object("error", "Password too short"));
            return;
        }
        
        store.updatePassword(userId, (String)newP);
        sendJson(exchange, 200, JsonHelper.object());
    }
}

class TodosHandler extends BaseHandler {
    public TodosHandler(DataStore store) { super(store); }
    public void handle(HttpExchange exchange) throws IOException {
        String method = exchange.getRequestMethod();
        String path = exchange.getRequestURI().getPath();
        
        if (path.equals("/todos")) {
            if (method.equals("GET")) {
                Integer userId = requireAuth(exchange);
                if (userId == null) return;
                List<Todo> todos = store.getTodosByUser(userId);
                List<String> jsonTodos = new ArrayList<>();
                for (Todo t : todos) jsonTodos.add(t.toJson());
                sendJson(exchange, 200, JsonHelper.array(jsonTodos));
            } else if (method.equals("POST")) {
                Integer userId = requireAuth(exchange);
                if (userId == null) return;
                String body = readBody(exchange);
                Map<String, Object> req = JsonHelper.parse(body);
                Object titleObj = req.get("title");
                
                if (!(titleObj instanceof String) || ((String)titleObj).isEmpty()) {
                    sendJson(exchange, 400, JsonHelper.object("error", "Title is required"));
                    return;
                }
                String title = (String) titleObj;
                String desc = req.containsKey("description") && req.get("description") instanceof String ? (String) req.get("description") : "";
                
                Todo todo = store.createTodo(userId, title, desc);
                sendJson(exchange, 201, todo.toJson());
            } else {
                sendJson(exchange, 405, JsonHelper.object("error", "Method not allowed"));
            }
        } else if (path.startsWith("/todos/")) {
            new TodoIdHandler(store).handle(exchange);
        } else {
            sendJson(exchange, 404, JsonHelper.object("error", "Not found"));
        }
    }
}

class TodoIdHandler extends BaseHandler {
    public TodoIdHandler(DataStore store) { super(store); }
    public void handle(HttpExchange exchange) throws IOException {
        String method = exchange.getRequestMethod();
        String path = exchange.getRequestURI().getPath();
        
        String[] parts = path.split("/");
        if (parts.length != 3) {
            sendJson(exchange, 404, JsonHelper.object("error", "Not found"));
            return;
        }
        
        int id;
        try {
            id = Integer.parseInt(parts[2]);
        } catch (NumberFormatException e) {
            sendJson(exchange, 404, JsonHelper.object("error", "Not found"));
            return;
        }
        
        Integer userId = requireAuth(exchange);
        if (userId == null) return;
        
        if (method.equals("GET")) {
            Todo todo = store.getTodo(id);
            if (todo == null || todo.userId != userId) {
                sendJson(exchange, 404, JsonHelper.object("error", "Todo not found"));
                return;
            }
            sendJson(exchange, 200, todo.toJson());
        } else if (method.equals("PUT")) {
            Todo todo = store.getTodo(id);
            if (todo == null || todo.userId != userId) {
                sendJson(exchange, 404, JsonHelper.object("error", "Todo not found"));
                return;
            }
            String body = readBody(exchange);
            Map<String, Object> req = JsonHelper.parse(body);
            
            String title = null;
            if (req.containsKey("title")) {
                if (!(req.get("title") instanceof String)) {
                    sendJson(exchange, 400, JsonHelper.object("error", "Title is required"));
                    return;
                }
                title = (String) req.get("title");
                if (title.isEmpty()) {
                    sendJson(exchange, 400, JsonHelper.object("error", "Title is required"));
                    return;
                }
            }
            
            String desc = null;
            if (req.containsKey("description")) {
                desc = req.get("description") instanceof String ? (String) req.get("description") : "";
            }
            
            Boolean completed = null;
            if (req.containsKey("completed")) {
                completed = req.get("completed") instanceof Boolean ? (Boolean) req.get("completed") : null;
            }
            
            Todo updated = store.updateTodo(id, userId, title, desc, completed);
            sendJson(exchange, 200, updated.toJson());
        } else if (method.equals("DELETE")) {
            boolean deleted = store.deleteTodo(id, userId);
            if (!deleted) {
                sendJson(exchange, 404, JsonHelper.object("error", "Todo not found"));
                return;
            }
            sendNoContent(exchange, 204);
        } else {
            sendJson(exchange, 405, JsonHelper.object("error", "Method not allowed"));
        }
    }
}

File: data_baseline/openrouter__qwen__qwen3.7-plus/rust/rep3/src/main.rs
use axum::{
    extract::{FromRef, FromRequestParts, Json, Path, State},
    http::{header, request::Parts, HeaderMap, HeaderValue, StatusCode},
    response::{IntoResponse, Response},
    routing::{get, post, put},
    Router,
};
use axum::async_trait;
use axum_extra::extract::CookieJar;
use chrono::Utc;
use clap::Parser;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use uuid::Uuid;

#[derive(Parser, Debug)]
#[command(name = "todo_app")]
#[command(about = "A REST API server for managing personal todo items")]
struct Args {
    #[arg(long)]
    port: u16,
}

#[derive(Clone)]
pub struct AppState {
    pub users: Arc<RwLock<HashMap<i32, User>>>,
    pub todos: Arc<RwLock<HashMap<i32, Todo>>>,
    pub sessions: Arc<RwLock<HashMap<String, i32>>>,
    pub next_user_id: Arc<RwLock<i32>>,
    pub next_todo_id: Arc<RwLock<i32>>,
}

pub struct User {
    pub id: i32,
    pub username: String,
    pub password: String,
}

#[derive(Serialize, Clone, Deserialize)]
pub struct Todo {
    pub id: i32,
    pub title: String,
    pub description: String,
    pub completed: bool,
    pub created_at: String,
    pub updated_at: String,
    #[serde(skip_serializing)]
    pub user_id: i32,
}

#[derive(Deserialize)]
pub struct RegisterRequest {
    pub username: String,
    pub password: String,
}

#[derive(Serialize)]
pub struct UserResponse {
    pub id: i32,
    pub username: String,
}

#[derive(Deserialize)]
pub struct LoginRequest {
    pub username: String,
    pub password: String,
}

#[derive(Deserialize)]
pub struct PasswordRequest {
    pub old_password: String,
    pub new_password: String,
}

#[derive(Deserialize)]
pub struct TodoRequest {
    pub title: Option<String>,
    pub description: Option<String>,
}

#[derive(Deserialize)]
pub struct TodoUpdateRequest {
    pub title: Option<String>,
    pub description: Option<String>,
    pub completed: Option<bool>,
}

pub enum AppError {
    BadRequest(String),
    NotFound(String),
    Unauthorized(String),
    Conflict(String),
}

impl IntoResponse for AppError {
    fn into_response(self) -> Response {
        let (status, message) = match self {
            AppError::BadRequest(msg) => (StatusCode::BAD_REQUEST, msg),
            AppError::NotFound(msg) => (StatusCode::NOT_FOUND, msg),
            AppError::Unauthorized(msg) => (StatusCode::UNAUTHORIZED, msg),
            AppError::Conflict(msg) => (StatusCode::CONFLICT, msg),
        };
        (status, axum::Json(serde_json::json!({"error": message}))).into_response()
    }
}

pub struct AuthUser(pub i32);

#[async_trait]
impl<S> FromRequestParts<S> for AuthUser
where
    S: Send + Sync,
    AppState: FromRef<S>,
{
    type Rejection = AppError;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        let app_state = AppState::from_ref(state);
        let cookie_jar = CookieJar::from_headers(&parts.headers);
        
        if let Some(cookie) = cookie_jar.get("session_id") {
            let sessions = app_state.sessions.read().await;
            if let Some(&user_id) = sessions.get(cookie.value()) {
                return Ok(AuthUser(user_id));
            }
        }
        Err(AppError::Unauthorized("Authentication required".to_string()))
    }
}

async fn register(
    State(state): State<AppState>,
    Json(req): Json<RegisterRequest>,
) -> impl IntoResponse {
    let username_re = Regex::new(r"^[a-zA-Z0-9_]+$").unwrap();
    if !username_re.is_match(&req.username) || req.username.len() < 3 || req.username.len() > 50 {
        return AppError::BadRequest("Invalid username".to_string()).into_response();
    }
    if req.password.len() < 8 {
        return AppError::BadRequest("Password too short".to_string()).into_response();
    }
    
    let mut users = state.users.write().await;
    for user in users.values() {
        if user.username == req.username {
            return AppError::Conflict("Username already exists".to_string()).into_response();
        }
    }
    
    let mut next_id = state.next_user_id.write().await;
    let id = *next_id;
    *next_id += 1;
    
    users.insert(id, User {
        id,
        username: req.username.clone(),
        password: req.password.clone(),
    });
    
    (StatusCode::CREATED, axum::Json(UserResponse { id, username: req.username })).into_response()
}

async fn login(
    State(state): State<AppState>,
    Json(req): Json<LoginRequest>,
) -> impl IntoResponse {
    let users = state.users.read().await;
    if let Some(user) = users.values().find(|u| u.username == req.username && u.password == req.password) {
        let session_id = Uuid::new_v4().to_string();
        state.sessions.write().await.insert(session_id.clone(), user.id);
        
        let mut headers = HeaderMap::new();
        let cookie_val = format!("session_id={}; Path=/; HttpOnly", session_id);
        headers.insert(header::SET_COOKIE, HeaderValue::from_str(&cookie_val).unwrap());
        
        (
            StatusCode::OK,
            headers,
            axum::Json(UserResponse { id: user.id, username: user.username.clone() })
        ).into_response()
    } else {
        AppError::Unauthorized("Invalid credentials".to_string()).into_response()
    }
}

async fn logout(
    AuthUser(_user_id): AuthUser,
    cookie_jar: CookieJar,
    State(state): State<AppState>,
) -> impl IntoResponse {
    if let Some(cookie) = cookie_jar.get("session_id") {
        let mut sessions = state.sessions.write().await;
        sessions.remove(cookie.value());
    }
    let cookie_jar = cookie_jar.remove("session_id");
    (StatusCode::OK, cookie_jar, axum::Json(serde_json::json!({}))).into_response()
}

async fn me(
    AuthUser(user_id): AuthUser,
    State(state): State<AppState>,
) -> impl IntoResponse {
    let users = state.users.read().await;
    if let Some(user) = users.get(&user_id) {
        (StatusCode::OK, axum::Json(UserResponse { id: user.id, username: user.username.clone() })).into_response()
    } else {
        AppError::Unauthorized("Authentication required".to_string()).into_response()
    }
}

async fn update_password(
    AuthUser(user_id): AuthUser,
    State(state): State<AppState>,
    Json(req): Json<PasswordRequest>,
) -> impl IntoResponse {
    if req.new_password.len() < 8 {
        return AppError::BadRequest("Password too short".to_string()).into_response();
    }
    
    let mut users = state.users.write().await;
    if let Some(user) = users.get_mut(&user_id) {
        if user.password != req.old_password {
            return AppError::Unauthorized("Invalid credentials".to_string()).into_response();
        }
        user.password = req.new_password;
        (StatusCode::OK, axum::Json(serde_json::json!({}))).into_response()
    } else {
        AppError::Unauthorized("Authentication required".to_string()).into_response()
    }
}

async fn list_todos(
    AuthUser(user_id): AuthUser,
    State(state): State<AppState>,
) -> impl IntoResponse {
    let todos = state.todos.read().await;
    let mut user_todos: Vec<Todo> = todos.values()
        .filter(|t| t.user_id == user_id)
        .cloned()
        .collect();
    user_todos.sort_by_key(|t| t.id);
    
    (StatusCode::OK, axum::Json(user_todos)).into_response()
}

async fn create_todo(
    AuthUser(user_id): AuthUser,
    State(state): State<AppState>,
    Json(req): Json<TodoRequest>,
) -> impl IntoResponse {
    let title = match req.title {
        Some(t) if !t.is_empty() => t,
        _ => return AppError::BadRequest("Title is required".to_string()).into_response(),
    };
    
    let description = req.description.unwrap_or_default();
    let now = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
    
    let mut next_id = state.next_todo_id.write().await;
    let id = *next_id;
    *next_id += 1;
    
    let todo = Todo {
        id,
        user_id,
        title,
        description,
        completed: false,
        created_at: now.clone(),
        updated_at: now,
    };
    
    state.todos.write().await.insert(id, todo.clone());
    
    (StatusCode::CREATED, axum::Json(todo)).into_response()
}

async fn get_todo(
    AuthUser(user_id): AuthUser,
    Path(todo_id): Path<i32>,
    State(state): State<AppState>,
) -> impl IntoResponse {
    let todos = state.todos.read().await;
    if let Some(todo) = todos.get(&todo_id) {
        if todo.user_id == user_id {
            (StatusCode::OK, axum::Json(todo.clone())).into_response()
        } else {
            AppError::NotFound("Todo not found".to_string()).into_response()
        }
    } else {
        AppError::NotFound("Todo not found".to_string()).into_response()
    }
}

async fn update_todo(
    AuthUser(user_id): AuthUser,
    Path(todo_id): Path<i32>,
    State(state): State<AppState>,
    Json(req): Json<TodoUpdateRequest>,
) -> impl IntoResponse {
    let mut todos = state.todos.write().await;
    if let Some(todo) = todos.get_mut(&todo_id) {
        if todo.user_id != user_id {
            return AppError::NotFound("Todo not found".to_string()).into_response();
        }
        
        if let Some(title) = req.title {
            if title.is_empty() {
                return AppError::BadRequest("Title is required".to_string()).into_response();
            }
            todo.title = title;
        }
        if let Some(description) = req.description {
            todo.description = description;
        }
        if let Some(completed) = req.completed {
            todo.completed = completed;
        }
        todo.updated_at = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
        
        (StatusCode::OK, axum::Json(todo.clone())).into_response()
    } else {
        AppError::NotFound("Todo not found".to_string()).into_response()
    }
}

async fn delete_todo(
    AuthUser(user_id): AuthUser,
    Path(todo_id): Path<i32>,
    State(state): State<AppState>,
) -> impl IntoResponse {
    let mut todos = state.todos.write().await;
    if let Some(todo) = todos.get(&todo_id) {
        if todo.user_id != user_id {
            return AppError::NotFound("Todo not found".to_string()).into_response();
        }
        todos.remove(&todo_id);
        StatusCode::NO_CONTENT.into_response()
    } else {
        AppError::NotFound("Todo not found".to_string()).into_response()
    }
}

#[tokio::main]
async fn main() {
    let args = Args::parse();
    let addr = format!("0.0.0.0:{}", args.port);
    println!("Listening on {}", addr);
    
    let state = AppState {
        users: Arc::new(RwLock::new(HashMap::new())),
        todos: Arc::new(RwLock::new(HashMap::new())),
        sessions: Arc::new(RwLock::new(HashMap::new())),
        next_user_id: Arc::new(RwLock::new(1)),
        next_todo_id: Arc::new(RwLock::new(1)),
    };

    let app = Router::new()
        .route("/register", post(register))
        .route("/login", post(login))
        .route("/logout", post(logout))
        .route("/me", get(me))
        .route("/password", put(update_password))
        .route("/todos", get(list_todos).post(create_todo))
        .route("/todos/:id", get(get_todo).put(update_todo).delete(axum::routing::delete(delete_todo)))
        .with_state(state);

    let listener = tokio::net::TcpListener::bind(&addr).await.unwrap();
    axum::serve(listener, app).await.unwrap();
}